src/MapIterableCapableTrait.php

```diff
@@ -0,0 +1,114 @@
+<?php
+
+namespace Dhii\Iterator;
+
+use Dhii\Invocation\Exception\InvocationExceptionInterface;
+use Dhii\Util\String\StringableInterface as Stringable;
+use InvalidArgumentException;
+use stdClass;
+use Traversable;
+
+trait MapIterableCapableTrait
+{
+    /**
+     * Invokes a callback for each element of the iterable.
+     *
+     * @since [*next-version*]
+     *
+     * @param array|stdClass|Traversable            $iterable The iterable to map.
+     * @param callable                              $callback
+     * @param Stringable|string|int|float|bool|null $start    The offset of the iteration, at which to start applying the callback.
+     *                                                        Iterations will still happen on all previous elements, but the callback will not be applied.
+     *                                                        Default: 0.
+     * @param Stringable|string|int|float|bool|null $count    The number  of invocations to make. Iteration will stop when this number is reached.
+     *                                                        Pass 0 (zero) to iterate until end.
+     *                                                        Default: 0.
+     * @param array|null                            $results  This will be filled with the results of the callback, in the same order, preserving keys.
```

Comment:
It's not documented that passing null does _not_ populate it with the results.

Reply:
Does not populate *what* with results?

Reply:
The `$results` reference.

The doc for the `$results` parameter says:

> This will be filled with the results of the callback, in the same order, preserving keys.

But if a `null` argument is given for this parameter, this does not happen based on the following code at line 56:

```
if (is_array($results)) {
    $results[$_k] = $result;
}
```

Reply:
Can neither populate `null`, or pass it by reference.

Reply:
Yes, but the method can initialize the array and populate it anyway.

```
if ($results === null) {
	$results = [];
}
```

It does not do this, so it should be documented IMO.

Reply:
It can initialize an array and populate it, but it will not be the reference passed, and therefore cannot be used outside.

Reply:
[Yes it can](https://3v4l.org/9PvUE).

Reply:
Good point.